<a href="https://colab.research.google.com/github/tutalae/New-York-Yellow-Taxi-Fare/blob/main/New_York_Yellow_Taxi_Fare.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [47]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from datetime import date
import datetime as dt
import calendar

In [74]:
url = 'https://drive.google.com/file/d/1QcKzIJlO6WRUH4IkrKAvDmBVEipXkeQb/view?usp=sharing'

path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]

data = pd.read_csv(path)

In [4]:
data.head()

VendorID tpep_pickup_datetime  ... DOLocationID  fare_amount
0       1.0  2020-01-01 00:28:15  ...          239          6.0
1       1.0  2020-01-01 00:35:39  ...          238          7.0
2       1.0  2020-01-01 00:47:41  ...          238          6.0
3       1.0  2020-01-01 00:55:23  ...          151          5.5
4       2.0  2020-01-01 00:01:58  ...          193          3.5

[5 rows x 7 columns]

In [5]:
data.describe()

VendorID  passenger_count  PULocationID  DOLocationID   fare_amount
count  1.227632e+06     1.227632e+06  1.243241e+06  1.243241e+06  1.243241e+06
mean   1.676771e+00     1.568692e+00  1.627453e+02  1.607467e+02  1.302977e+01
std    4.677094e-01     1.181809e+00  6.610786e+01  7.098176e+01  1.293962e+01
min    1.000000e+00     0.000000e+00  1.000000e+00  1.000000e+00 -1.238000e+03
25%    1.000000e+00     1.000000e+00  1.250000e+02  1.070000e+02  6.000000e+00
50%    2.000000e+00     1.000000e+00  1.610000e+02  1.620000e+02  9.000000e+00
75%    2.000000e+00     2.000000e+00  2.330000e+02  2.330000e+02  1.400000e+01
max    2.000000e+00     9.000000e+00  2.650000e+02  2.650000e+02  1.238000e+03

- Maximum passenger count for a trip is 9
- Maximum fare amount for a trip is 1238 dollar 
- Average fare amount for trips is 13.03 dollars

In [25]:
data["PULocationID"].value_counts()

132    53790
237    52055
161    49558
236    48450
186    44809
       ...  
206        1
245        1
111        1
30         1
2          1
Name: PULocationID, Length: 255, dtype: int64

In [24]:
data["DOLocationID"].value_counts()

236    52377
237    46683
161    44751
230    39926
170    36256
       ...  
187        7
204        7
99         3
2          3
104        1
Name: DOLocationID, Length: 261, dtype: int64

- The count of distinct pickup location id is less than the count of drop off location id

- Most frequently used pick up location id is 132 and least frequent drop off location id is 104

Busiest day

In [58]:
data["tpep_pickup_datetime"] = pd.to_datetime(data["tpep_pickup_datetime"])
data["tpep_dropoff_datetime"] = pd.to_datetime(data["tpep_dropoff_datetime"])

data["tpep_pickup_day"] = data["tpep_dropoff_datetime"] .dt.day_name()
data["tpep_dropoff_day"] = data["tpep_dropoff_datetime"] .dt.day_name()

data.head(3)

VendorID tpep_pickup_datetime  ... tpep_pickup_day  tpep_dropoff_day
0       1.0  2020-01-01 00:28:15  ...       Wednesday         Wednesday
1       1.0  2020-01-01 00:35:39  ...       Wednesday         Wednesday
2       1.0  2020-01-01 00:47:41  ...       Wednesday         Wednesday

[3 rows x 9 columns]

In [60]:
data["tpep_pickup_day"].value_counts()

Tuesday      201286
Friday       182810
Saturday     182314
Monday       179696
Wednesday    169470
Sunday       165612
Thursday     162053
Name: tpep_pickup_day, dtype: int64

In [61]:
data["tpep_dropoff_day"].value_counts()

Tuesday      201286
Friday       182810
Saturday     182314
Monday       179696
Wednesday    169470
Sunday       165612
Thursday     162053
Name: tpep_dropoff_day, dtype: int64

- Tuesday is the busiest day in terms of passenger pickups.

- Tuesday is the busiest day in terms of passenger drop-offs.

Create a subset of the data by extracting those observations where:
 - vendor ID is 2
 - fare amount is less than the mean 'fare amount'
 - 'passenger count' is greater than 5.

In [82]:
new_df = data[data["VendorID"]==2][data["fare_amount"]<1.302977e+01][data["passenger_count"]>5]
new_df.tail()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  """Entry point for launching an IPython kernel.


VendorID tpep_pickup_datetime  ... DOLocationID  fare_amount
1227569       2.0  2020-01-07 20:28:35  ...          170          7.0
1227576       2.0  2020-01-07 23:57:24  ...          186          8.0
1227577       2.0  2020-01-07 23:55:28  ...          163          6.0
1227592       2.0  2020-01-07 23:58:26  ...           79          9.5
1227611       2.0  2020-01-07 23:59:13  ...          238          6.5

[5 rows x 7 columns]

In [83]:
new_df.info

<bound method DataFrame.info of          VendorID tpep_pickup_datetime  ... DOLocationID  fare_amount
195           2.0  2020-01-01 00:38:43  ...          170         12.0
218           2.0  2020-01-01 00:24:07  ...          260          5.5
219           2.0  2020-01-01 00:39:43  ...          260          7.0
220           2.0  2020-01-01 00:48:29  ...          129          6.5
453           2.0  2020-01-01 00:47:13  ...          140          7.5
...           ...                  ...  ...          ...          ...
1227569       2.0  2020-01-07 20:28:35  ...          170          7.0
1227576       2.0  2020-01-07 23:57:24  ...          186          8.0
1227577       2.0  2020-01-07 23:55:28  ...          163          6.0
1227592       2.0  2020-01-07 23:58:26  ...           79          9.5
1227611       2.0  2020-01-07 23:59:13  ...          238          6.5

[19443 rows x 7 columns]>

In [81]:
new_df.isna().sum()

VendorID                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
PULocationID             0
DOLocationID             0
fare_amount              0
dtype: int64

- The total number of observations in 'new_df' is 19443 and there are no missing values in any column.

In [84]:
new_df.describe()

VendorID  passenger_count  PULocationID  DOLocationID   fare_amount
count   19443.0     19443.000000  19443.000000  19443.000000  19443.000000
mean        2.0         6.000257    165.997994    165.288381      7.632995
std         0.0         0.025857     66.614747     68.441004      2.623450
min         2.0         6.000000      4.000000      4.000000    -65.000000
25%         2.0         6.000000    114.000000    114.000000      5.500000
50%         2.0         6.000000    162.000000    162.000000      7.500000
75%         2.0         6.000000    234.000000    234.000000      9.500000
max         2.0         9.000000    265.000000    265.000000     13.000000

The mean value of 'fare amount' of new_df is 7.63 (rounded to two decimal points) and the average passenger count is approximately 6 (ignore the values after decimal)

In [85]:
new_df["tpep_pickup_datetime"] = pd.to_datetime(new_df["tpep_pickup_datetime"])
new_df["tpep_dropoff_datetime"] = pd.to_datetime(new_df["tpep_dropoff_datetime"])

new_df["tpep_pickup_day"] = new_df["tpep_dropoff_datetime"] .dt.day_name()
new_df["tpep_dropoff_day"] = new_df["tpep_dropoff_datetime"] .dt.day_name()

new_df.head(3)

VendorID tpep_pickup_datetime  ... tpep_pickup_day  tpep_dropoff_day
195       2.0  2020-01-01 00:38:43  ...       Wednesday         Wednesday
218       2.0  2020-01-01 00:24:07  ...       Wednesday         Wednesday
219       2.0  2020-01-01 00:39:43  ...       Wednesday         Wednesday

[3 rows x 9 columns]

In [89]:
new_df["tpep_dropoff_day"].value_counts()

Saturday     3072
Tuesday      3062
Friday       2895
Monday       2663
Wednesday    2628
Thursday     2618
Sunday       2505
Name: tpep_dropoff_day, dtype: int64

Saturday is the busiest day in terms of most number of pickups while Sunday has the least number of passenger drop-offs.

In [90]:
For new_df, the median fare amount is less than the mean fare amount

SyntaxError: ignored